File: packages/shared-types/src/team.ts
import type { Timestamp } from './index'
import type { TeamTagRef } from './tag'

/**
 * Document `/teams/{teamId}` — équipe du club.
 * Voir docs/firebase.md (section /teams/{teamId}).
 *
 * Les teams persistent cross-saisons via `activeSeasonIds[]`. La cotisation
 * de l'équipe est référencée via `cotisationId` (cf. /cotisations).
 *
 * Note: pour l'instant seuls les champs nécessaires au Dashboard sont
 * exposés strictement (`id`, `name`, `categoryId`). Les autres seront ajoutés
 * au fil du dev des écrans Teams / Season planning assistant.
 *
 * `categoryId` est une référence vers `/categories/{categoryId}` (référentiel
 * éditable par l'admin). Le libellé et la tranche d'âge sont résolus à la
 * lecture côté repo — pas de dénormalisation.
 */
export type TeamGender = 'M' | 'F' | 'mixed'

export type TeamRegistrationStatus = 'open' | 'conditional' | 'closed'

export interface TeamSchedulingPreferredDay {
  dayOfWeek: number
  priority: number
}

export interface TeamCoachAvailability {
  coachMemberId: string
  unavailableDays: number[]
  unavailableSlots: string[]
}

export interface TeamSchedulingConstraints {
  preferredDays: TeamSchedulingPreferredDay[]
  /** "HH:MM" */
  maxStartTime: string
  minHoursBetweenSlots: number
  trainingsPerWeek: number
  anticipatedMatches: number
  coachAvailability: TeamCoachAvailability[]
}

/**
 * Lien entre une équipe court-base et une compétition Basketplan (Swiss
 * Basketball / ORCA Systems). Une même équipe peut être inscrite dans
 * **plusieurs** compétitions (championnat + coupe, ou plusieurs fédérations
 * en parallèle) — d'où le tableau `basketplanLinks[]` ci-dessous.
 *
 * Voir `docs/basketplan-integration.md` § 4.1 pour la spec complète et
 * `docs/chantier-basketplan.md` (PR 1) pour le plan d'exécution. Le lien
 * porte des **caches** d'affichage (`federationCode`, `leagueHoldingName`,
 * `season`, `teamNameInLeague`) résolus côté serveur au moment du linkage,
 * pour éviter un re-fetch Basketplan à chaque rendu de la liste.
 *
 * Identifiants externes :
 *  - `federationId` : id numérique de la fédération Basketplan (ex. 9 = AFBB).
 *  - `leagueHoldingId` : id de la compétition (championnat / coupe) saison-précis.
 *  - `teamIdInLeague` : id Basketplan de l'équipe DANS cette compétition
 *    (différent d'une compétition à l'autre, même pour la "même" équipe club).
 *
 * Cycle de vie :
 *  - Créé par les callables `linkTeamToBasketplan` (résout les caches + pose
 *    `id` uuid, `addedAt`, `addedBy`).
 *  - Toggleable via `toggleTeamBasketplanLink` (`active: false` = pause sans
 *    perte). Retiré via `unlinkTeamBasketplan`.
 *  - Les `leagueHoldingId` changent à chaque saison Basketplan : à la fin
 *    d'une saison, prévoir un mécanisme de "renouveler les liens" (cf.
 *    brief § 7.2).
 */
export interface BasketplanCompetitionLink {
  /** UUID local, généré côté Cloud Function via `crypto.randomUUID()`. */
  id: string
  /** Id numérique de la fédération Basketplan (ex. 9 = AFBB). */
  federationId: number
  /** Code court de la fédération (cache pour affichage rapide, ex. "AFBB"). */
  federationCode: string
  /** Id Basketplan de la compétition (championnat ou coupe) — saison-précis. */
  leagueHoldingId: number
  /** Nom complet de la compétition (cache pour affichage). */
  leagueHoldingName: string
  /** Saison extraite du nom (ex. "25/26"). Cache pour groupement UI. */
  season: string
  /** Id Basketplan de l'équipe **dans cette ligue** (≠ id "team club" global). */
  teamIdInLeague: number
  /** Nom de l'équipe tel qu'inscrit dans cette ligue (cache, ex. "Marly Basket 2LM"). */
  teamNameInLeague: string
  /** Pause sans suppression — sync ignore les liens inactifs. */
  active: boolean
  addedAt: Timestamp
  /** uid du caller (admin ou coach) au moment du linkage. */
  addedBy: string
}

export interface TeamData {
  /** Ex. "U20F" */
  name: string
  /** Référence vers `/categories/{categoryId}`. */
  categoryId: string
  gender: TeamGender
  /** memberIds des coachs (cumulable avec /users.teamIds côté coach scope). */
  coachIds: string[]
  /** memberIds des joueurs. */
  playerIds: string[]
  activeSeasonIds: string[]
  /**
   * Référence vers `/cotisations/{id}` — montant et description résolus à la
   * lecture (pas dénormalisé). Cf. docs/main.md → 'Cotisations'.
   */
  cotisationId: string
  schedulingConstraints: TeamSchedulingConstraints
  /**
   * Tags attachés à l'équipe (référence + flag d'affichage par-équipe).
   * Cf. /tags référentiel et docs/main.md ("Tags d'équipes").
   */
  tags: TeamTagRef[]
  /**
   * Statut d'ouverture aux nouvelles inscriptions (saison courante).
   * Affiché publiquement dans l'app register (TeamPicker §4.5 du brief).
   */
  registrationStatus: TeamRegistrationStatus
  /**
   * Manuel d'inscription affiché en branche "équipe ouverte" (markdown court).
   * Écrit par le coach ou l'admin depuis l'app web.
   */
  openHandbook: string
  /**
   * Description des conditions affichée en branche "équipe sous conditions"
   * (markdown court).
   */
  conditionalDescription: string
  /**
   * Liste de critères tagués pour les équipes sous conditions (chips d'affichage).
   */
  conditionalCriteria: string[]
  /**
   * Accroche courte affichée sur la fiche publique de l'équipe (app register).
   * `null` = pas d'accroche, on affiche le nom de l'équipe seul.
   */
  publicTagline: string | null
  /**
   * memberId du coach affiché comme "head coach" sur la fiche publique
   * (quand `coachIds` en contient plusieurs). `null` = premier de la liste utilisé.
   */
  publicHeadCoachMemberId: string | null
  active: boolean
  createdAt: Timestamp
  /**
   * Liens Basketplan (1→N) — voir `BasketplanCompetitionLink` au-dessus et
   * `docs/basketplan-integration.md` § 4.1. `undefined` ou `[]` = équipe
   * jamais liée. Toutes les mutations passent par les callables
   * `linkTeamToBasketplan` / `unlinkTeamBasketplan` / `toggleTeamBasketplanLink`
   * (Admin SDK — `/teams` reste write admin-only côté rules).
   */
  basketplanLinks?: BasketplanCompetitionLink[]
}

export type Team = TeamData & { id: string }

File: apps/mobile/lib/models/team.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../core/firestore_converters.dart';
import 'enums.dart';

/// Référence à un tag d'équipe (`/teams.tags[]`).
class TeamTagRef {
  const TeamTagRef({required this.tagId, required this.display});

  final String tagId;

  /// Flag d'affichage propre à cette équipe.
  final bool display;

  factory TeamTagRef.fromMap(Map<String, dynamic> map) => TeamTagRef(
        tagId: FsConvert.str(map['tagId']),
        display: FsConvert.boolean(map['display']),
      );
}

/// Document `/teams/{teamId}` — équipe du club.
///
/// `categoryId` et `cotisationId` sont des **références** : le libellé / prix
/// sont résolus à la lecture (pas dénormalisés). Seul un sous-ensemble des
/// champs du schéma est modélisé ici — les contraintes de planning et les
/// champs publics ne sont pas requis par l'app mobile officiel/coach.
class Team {
  const Team({
    required this.id,
    required this.name,
    required this.categoryId,
    required this.gender,
    required this.coachIds,
    required this.playerIds,
    required this.activeSeasonIds,
    required this.cotisationId,
    required this.tags,
    required this.registrationStatus,
    required this.active,
    required this.createdAt,
  });

  final String id;
  final String name;

  /// Référence vers `/categories/{categoryId}`.
  final String categoryId;
  final TeamGender gender;

  /// memberIds des coachs.
  final List<String> coachIds;

  /// memberIds des joueurs.
  final List<String> playerIds;
  final List<String> activeSeasonIds;

  /// Référence vers `/cotisations/{id}`.
  final String cotisationId;
  final List<TeamTagRef> tags;
  final TeamRegistrationStatus registrationStatus;
  final bool active;
  final DateTime createdAt;

  factory Team.fromMap(String id, Map<String, dynamic> map) => Team(
        id: id,
        name: FsConvert.str(map['name']),
        categoryId: FsConvert.str(map['categoryId']),
        gender: TeamGender.fromWire(map['gender']),
        coachIds: FsConvert.stringList(map['coachIds']),
        playerIds: FsConvert.stringList(map['playerIds']),
        activeSeasonIds: FsConvert.stringList(map['activeSeasonIds']),
        cotisationId: FsConvert.str(map['cotisationId']),
        tags: FsConvert.mapList(map['tags'])
            .map(TeamTagRef.fromMap)
            .toList(growable: false),
        registrationStatus:
            TeamRegistrationStatus.fromWire(map['registrationStatus']),
        active: FsConvert.boolean(map['active'], true),
        createdAt: FsConvert.toDateTimeOr(map['createdAt']),
      );

  factory Team.fromFirestore(
    DocumentSnapshot<Map<String, dynamic>> snapshot,
  ) =>
      Team.fromMap(snapshot.id, FsConvert.requireData(snapshot));

  Team copyWith({
    String? name,
    String? categoryId,
    TeamGender? gender,
    List<String>? coachIds,
    List<String>? playerIds,
    List<String>? activeSeasonIds,
    String? cotisationId,
    List<TeamTagRef>? tags,
    TeamRegistrationStatus? registrationStatus,
    bool? active,
    DateTime? createdAt,
  }) =>
      Team(
        id: id,
        name: name ?? this.name,
        categoryId: categoryId ?? this.categoryId,
        gender: gender ?? this.gender,
        coachIds: coachIds ?? this.coachIds,
        playerIds: playerIds ?? this.playerIds,
        activeSeasonIds: activeSeasonIds ?? this.activeSeasonIds,
        cotisationId: cotisationId ?? this.cotisationId,
        tags: tags ?? this.tags,
        registrationStatus: registrationStatus ?? this.registrationStatus,
        active: active ?? this.active,
        createdAt: createdAt ?? this.createdAt,
      );
}
